src/ai/defaultai.cc

```diff
@@ -1628,39 +1781,39 @@ void DefaultAI::update_buildable_field(BuildableField& field) {
 		   map.find_bobs(game(), Widelands::Area<Widelands::FCoords>(field.coords, kProductionArea),
 		                 nullptr, Widelands::FindBobCritter());
 
-		// Rocks are not renewable, we will count them only if previous state is nonzero
-		if (field.rocks_nearby > 0) {
-			field.rocks_nearby = map.find_immovables(
-			   game(),
-			   Widelands::Area<Widelands::FCoords>(map.get_fcoords(field.coords), kProductionArea),
-			   nullptr,
-			   Widelands::FindImmovableAttribute(
-			      Widelands::MapObjectDescr::get_attribute_id("rocks")));
+		// Counting trees, rocks, berry bushes nearby
+		for (const auto& attribute_category : buildings_immovable_attributes_) {
+			for (const ImmovableAttribute& attribute_info : attribute_category.second) {
+				// Rocks are not renewable, we will count them only if they were never counted or
+				// previous state is nonzero. This is done to save some work for AI (CPU utilization).
+				if (attribute_info.building_attribute == BuildingAttribute::kNeedsRocks) {
+					auto needs_rocks_it =
+					   field.immovables_by_attribute_nearby.find(BuildingAttribute::kNeedsRocks);
+					if (needs_rocks_it != field.immovables_by_attribute_nearby.end() &&
+					    needs_rocks_it->second == 0) {
+						continue;
+					}
+				}
+				uint8_t amount = map.find_immovables(
+				   game(),
+				   Widelands::Area<Widelands::FCoords>(map.get_fcoords(field.coords), kProductionArea),
+				   nullptr, Widelands::FindImmovableAttribute(attribute_category.first));
+
+				// adding 2 if rocks found
+				if (attribute_info.building_attribute == BuildingAttribute::kNeedsRocks) {
```

Comment:
I beleive this line is always adding 2 regardless whether there has been found some rocks in the vicinity this should be dependent on amount found as it was in previous code.